### PR TITLE
net: l2: ieee802154: standardize RSSI value

### LINKED ISSUE #58494: drivers: ieee802154: RSSI value converted inconsistently across drivers
**Describe the bug**

The RSSI value in net_pkt (net_pkt_cb_ieee802154.rssi) is used inconsistently across drivers.

Example 1: At least one driver (nRF5) casts the signed RSSI (dBm?) value returned from the HW to the unsigned net_pkt_cb_ieee802154.rssi value without conversion.

Example 2: The cc13/26xx drivers convert the signed RSSI dBm value returned from the HW to an unsigned unitless value evenly spread to the 0-255 uint8 range of net_pkt_cb_ieee802154.rssi.

What is the intended use and range of this value? Depending on the answer one of the two implementations is probably wrong.

**To Reproduce**

n/a

**Expected behavior**

As proposed by @nandojve below, the RSSI value in net_pkt could be interpreted as defined in the IEEE 802.15.4 standard. This makes even more sense as the attribute is part of an IEEE 802.15.4-specific control block. This means that we don't have to change the API of net_pkt at all but only document the expected value range plus provide some utility functions for dBm-to-macRSSI and macRSSI-to-dBm conversion.

> 6.16.2.8 Received Signal Strength Indicator (RSSI)
> ...
> RSSI is represented as one octet of integer as shown in Table 8-108; therefore, the minimum and maximum values
> are 0 (–174 dBm) and 254 (80 dBm), respectively. 255 is reserved.

I can provide fixes for net core and at least the nRF5 and CC13/26xx drivers if we agree on this intended behavior.

**Impact**

RSSI value is not comparable between different driver implementations and its physical interpretation unclear or ambivalent. Therefore it is probably not too useful right now.

**Logs and console output**

n/a

**Environment (please complete the following information):**

n/a

### PR DESCRIPTION
**Pull request by ghost**
The RSSI value in net_pkt (net_pkt_cb_ieee802154.rssi) was used inconsistently across drivers. Some drivers did cast a signed dBm value directly to net_pkt's unsigned byte value. Others were assigning the negative value of the signed dBm value and again others were offsetting and stretching the signed dBm value linearly onto the full unsigned byte range.

This change standardizes net_pkt's rssi attribute to represent RSSI on the RX path as an unsigned integer ranging from 0 (–174 dBm) to 254 (80 dBm) and lets 255 represent an "unknown RSSI" (IEEE 802.15.4-2020, section 6.16.2.8). On the TX path the rssi attribute will always be zero. Out-of-range values will be truncated to max/min values.

The change also introduces conversion functions to and from signed dBm values and introduces these consistently to all existing call sites. The "unknown RSSI" value is represented as INT16_MIN in this case.

In some cases drivers had to be changed to calculate dBm values from internal hardware specific representations.

The conversion functions are fully covered by unit tests.

Fixes: #58494

### CODE REVIEW

File: include/zephyr/net/ieee802154_pkt.h [hunk 1]
```diff
@@ -82,16 +95,89 @@ static inline void net_pkt_set_ieee802154_lqi(struct net_pkt *pkt, uint8_t lqi)
...
+ *            RSSI.
+ * @param rssi represented in dBm. Set to the special value
+ *             INT16_MIN if an RSSI value is not available for
+ *             this packet. Values above 80 dBm will be mapped to
+ *             80 dBm, values below -174 dBm will be mapped to
+ *             -174 dBm.
+ */
+static inline void net_pkt_set_ieee802154_rssi_dbm(struct net_pkt *pkt, int16_t rssi)
+{
+	if (likely(rssi >= -174 && rssi <= 80)) {
```
Comment (cfriedt) [inline include/zephyr/net/ieee802154_pkt.h:159]: Can you please `#define` physical layer constants for `-174` and `80` and use those here?

File: include/zephyr/net/ieee802154_pkt.h [hunk 2]
```diff
@@ -82,16 +95,89 @@ static inline void net_pkt_set_ieee802154_lqi(struct net_pkt *pkt, uint8_t lqi)
...
+		int16_t unsigned_rssi = rssi + 174;
+
+		net_pkt_cb_ieee802154(pkt)->rssi = unsigned_rssi;
+		return;
+	} else if (rssi == INT16_MIN) {
+		net_pkt_cb_ieee802154(pkt)->rssi = UINT8_MAX;
```
Comment (cfriedt) [inline include/zephyr/net/ieee802154_pkt.h:165]: Can you please also `#define` a physical layer constant for `UINT8_MAX`?

I'm also OK if this is defined as something like `((int8_t)-1)`, but I guess the spec says `0xff`?

File: include/zephyr/net/ieee802154_pkt.h [hunk 3]
```diff
@@ -82,16 +95,89 @@ static inline void net_pkt_set_ieee802154_lqi(struct net_pkt *pkt, uint8_t lqi)
...
+		return;
+	} else if (rssi > 80) {
+		net_pkt_cb_ieee802154(pkt)->rssi = UINT8_MAX - 1;
+		return;
+	}
+
+	/* As the above rule covers the whole int16_t range,
+	 * this code should never be reached.
+	 */
+	__ASSERT_NO_MSG(false);
```
Comment (cfriedt) [inline include/zephyr/net/ieee802154_pkt.h:178]:
```suggestion
	CODE_UNREACHABLE;
```

File: tests/net/ieee802154/l2/src/ieee802154_test.c [hunk 4]
```diff
@@ -535,6 +535,66 @@ static bool test_packet_cloning_with_cb(void)
...
+		NET_ERR("*** No pkt to allocate\n");
+		return false;
+	}
+
+	/* Test setting/getting of unsigned RSSI. */
+	net_pkt_set_ieee802154_rssi(pkt, 50U);
+	zassert_equal(net_pkt_ieee802154_rssi(pkt), 50U);
+
+	/* Test setting/getting of signed RSSI (in range). */
+	net_pkt_set_ieee802154_rssi_dbm(pkt, -174);
```
Comment (cfriedt) [inline tests/net/ieee802154/l2/src/ieee802154_test.c:557]: Looks great - thanks for adding to the testsuite :-)

The only suggestion I have is to use `#define`ed constant suggested earlier.

**Review by cfriedt — CHANGES_REQUESTED** [review]
Just minor changes requested. Overall - wow - thanks @fgrandel for your work here. I think this is the sort of consistence we require for interfacing with the physical layer.

A suggestion as well, can you please add your GitHub UserID (and mine as well for that matter) under `collaborators` for `Drivers: IEEE 802.15.4`? As long as there are no objections.

File: include/zephyr/net/ieee802154_pkt.h (hunk 1 above)
Comment (fgrandel) [reply include/zephyr/net/ieee802154_pkt.h]: Fixed.

File: include/zephyr/net/ieee802154_pkt.h (hunk 2 above)
Comment (fgrandel) [reply include/zephyr/net/ieee802154_pkt.h]: Fixed.

File: include/zephyr/net/ieee802154_pkt.h (hunk 3 above)
Comment (fgrandel) [reply include/zephyr/net/ieee802154_pkt.h]: Fixed.

File: tests/net/ieee802154/l2/src/ieee802154_test.c (hunk 4 above)
Comment (fgrandel) [reply tests/net/ieee802154/l2/src/ieee802154_test.c]: Fixed.

**Comment by fgrandel** [comment]
> A suggestion as well, can you please add your GitHub UserID (and mine as well for that matter) under `collaborators` for `Drivers: IEEE 802.15.4`? As long as there are no objections.

Added a (tentative) commit as suggested by @cfriedt . Maintainer/reviewers - pls object and I'l remove the commit again.

**Review by cfriedt — DISMISSED** [review]
LGTM in terms of common rssi, cc13xx_cc26xx, and MAINTAINERS changes

**Comment by cfriedt** [comment]
Just adding a DNM for now until `v3.4.0` is tagged

**Comment by fgrandel** [comment]
@cfriedt Inspired by your comments also fixed some magic RSSI values in CCxxxx drivers.

**Review by cfriedt — APPROVED** [review]

**Review by jciupis — APPROVED** [review]

**Review by rlubos — APPROVED** [review]
LGTM

**Review by nandojve — APPROVED** [review]

**Pull request merged**